lib/executor/shared_iterations.go

```diff
@@ -215,17 +215,20 @@ func (si SharedIterations) Run(ctx context.Context, out chan<- stats.SampleConta
 	runIteration := getIterationRunner(si.executionState, si.logger)
 
 	attemptedIters := new(uint64)
+
+	activationParams := getVUActivationParams(maxDurationCtx, si.config.BaseConfig,
+		func(u lib.InitializedVU) {
+			si.executionState.ReturnVU(u, true)
+			activeVUs.Done()
+		})
 	handleVU := func(initVU lib.InitializedVU) {
 		ctx, cancel := context.WithCancel(maxDurationCtx)
 		defer cancel()
 
-		vu := initVU.Activate(&lib.VUActivationParams{
-			RunContext: ctx,
-			DeactivateCallback: func() {
-				si.executionState.ReturnVU(initVU, true)
-				activeVUs.Done()
-			},
-		})
+		newParams := *activationParams
+		newParams.RunContext = ctx
```

Comment:
Wondering if these micro-optimizations are worth it. It probably saves a few ns, but calling `getVUActivationParams()` for every VU would be cleaner and less error-prone. Especially since `getVUActivationParams()` can probably be substantially optimized by being a private method of `BaseContext` that doesn't call any other functions and just uses the struct properties directly. Anyway, not a big deal, the current approach also LGTM.

Reply:
> Especially since getVUActivationParams() can probably be substantially optimized by being a private method of `BaseContext` that doesn't call any other functions and just uses the struct properties directly.

You mean a method on `BaseConfig`? Could be, but not sure I want it to meddle with VU stuff. I was considering dropping the helper function altogether, since it's really a `VUActivationParams` initializer, and doesn't help much. :) But to avoid this `RunContext` mess, I'll just call it for every VU instead, thanks.

Reply:
> Could be, but not sure I want it to meddle with VU stuff. I
:+1: , though it already meddles with them - that's where the `exec`, `env`, `tags` live after all. And looking at it again, it doesn't actually need to be a method, considering that the fields are public.

> But to avoid this RunContext mess, I'll just call it for every VU instead, thanks.
Only where it makes sense - IIRC some executors can happily reuse a single `VUActivationParams` object

Reply:
The quoting got messed up there :smile:, but I think we're on the same page. See 25a26ab6.